### PR TITLE
Travis CI Windows build

### PR DESCRIPTION
**Pull request by stukowski**
Hi!

This pull request provides the necessary changes to make Travis CI successfully build and test the KIM API on the Win32 platform. 

- Run the build script in the MinGW environment instead of MSYS2.
- Fix a crash in `Path::HomePath()` if `HOMEDRIVE`/`HOMEPATH` environment variables are not defined. That's the case in the Travis CI Windows build environment.
- Implement `PrivateGetORIGIN()` using Win32 API.
- Make sure auto-generated `kim-api-collections-management` wrapper script contains the correct relative path to `collections-info` executable on Windows.
- Avoid running `kim-api-collections-management install ...` from test script, which is a command not yet supported on Windows.

I'm sorry for the many Git commits in this pull request. It took some trial and error to get this to work.

### CODE REVIEW

File: scripts/ci-build-and-test [hunk 1]
```diff
@@ -83,12 +83,15 @@ printf -- "--- remaining installed files ---\n" &&
   find "${PWD}/destdir" &&
   printf -- "---------------------------------\n"
 set -xv
-dir="${INSTALL_PREFIX}"
-while test x"" != x"${dir}"; do
-  rmdir "${PWD}/destdir${dir}" || exit $?
-  dir=`printf -- "${dir}" | sed -e 's|\(.*\)/.*|\1|'`
-done
-rmdir "${PWD}/destdir" || exit $?
+# Skipping the following statements on Windows, because they fail in the mingw64 environment:
```
Comment (ellio167) [inline scripts/ci-build-and-test:86]: Interesting; why does it fail?  Is there a missing command or non-posix behavior?

Comment (stukowski) [reply scripts/ci-build-and-test:86]: Actually I don't know. It silently failed, and I didn't see any error messages in the Travis CI logs, unfortunately. Here is a log from a pipeline run where it got past the `${MAKE} install` line but then died before reaching the next `cmake` command:

https://travis-ci.com/github/stukowski/kim-api/builds/199660794

I couldn't tell what caused the failure, mainly because I still got to understand the purpose of that part of the test script. Could you please tell me what it is supposed to do? I will then take another look at it. Perhaps I can get it to work in the MinGW environment.

Comment (ellio167) [reply scripts/ci-build-and-test:86]: Right!  It is testing that nothing "extra" is being installed by the "make install" command and/or that everything is installed in the expected places.  It's not very complete for this purpose, but it tries to identify gross problems.

From the travis log you point to I see that the installed files have not been deleted by lines 78-80.  Probably this is due to the `INSTALL_PREFIX` variable having the windows drive name in it?

Comment (ellio167) [reply scripts/ci-build-and-test:86]: If we can't fix this then, I would put the `if` starting at line 76

Comment (stukowski) [reply scripts/ci-build-and-test:86]: I'll take another look at this later today. [CMake docs](https://cmake.org/cmake/help/latest/envvar/DESTDIR.html) say that the ` make DESTDIR` mechanism shouldn't be used on Windows, like you say (this might not apply to MinGW though). Perhaps it would be better to use the `cmake --install . --prefix` instead. I will do some local testing first before revising the CI script.

**Comment by stukowski** [comment]
After a couple of more fixes, I now got (almost) the entire CI test script to work in the Windows/MinGW environment. In particular:

- `make install DESTDIR=...` works 
- `kim-api-collections-management install ...` works 
- `kim-api-simulator-model` works 
- Running the commands from the `INSTALL` documentation file works.

However, the process gets stuck when it comes to downloading a driver from openkim.org, see the Travis CI log (no error, process doesn't terminate). Any idea why that happens? Is this even supposed to work on Windows? I'm not sure.

https://travis-ci.com/github/stukowski/kim-api/builds/200086002

**Comment by ellio167** [comment]
Once in a while the download fails for some reason.  I would re-trigger the ci build and see if it succeeds.

Great that you have it all working!

File: scripts/ci-build-and-test [hunk 2]
```diff
@@ -75,6 +75,10 @@ fi
 
 mkdir ${PWD}/destdir
 ${MAKE} install DESTDIR=${PWD}/destdir || exit $?
+# Query value of ${CMAKE_INSTALL_PREFIX} from CMake cache (https://stackoverflow.com/questions/8474753/how-to-get-a-cmake-variable-from-the-command-line)
+# Remove drive letter, which might precede the path on Windows (MinGW) platform.
+INSTALL_PREFIX=`cmake -L -N . | grep CMAKE_INSTALL_PREFIX | cut -d "=" -f2 | cut -d ":" -f2`
```
Comment (ellio167) [inline scripts/ci-build-and-test:80]: Why do you overwrite the value of `INSTALL_PREFIX` here?

File: examples/simulators/utility_forces_numer_deriv/CMakeLists.txt [hunk 3]
```diff
@@ -35,7 +35,7 @@
 
 cmake_minimum_required(VERSION 3.10)
 
-project(utility_forces_numer_deriv VERSION 1.0.0 LANGUAGES Fortran)
+project(utility_forces_numer_deriv VERSION 1.0.0 LANGUAGES Fortran CXX)
```
Comment (ellio167) [inline examples/simulators/utility_forces_numer_deriv/CMakeLists.txt:38]: Why is CXX needed here?

File: scripts/ci-build-and-test [hunk 4]
```diff
@@ -109,28 +113,29 @@ ccache -s
...
-printf "calling ${PWD}/test-instal-relocatedl/bin/kim-api-activate\n"
+printf "calling ${PWD}/test-install-relocated/bin/kim-api-activate\n"
 source "${PWD}/test-install-relocated/bin/kim-api-activate" || exit $?
 set -xv
 kim-api-collections-management install system `find ../examples/model-drivers -mindepth 1 -maxdepth 1 -type d` &&
   kim-api-collections-management install system `find ../examples/portable-models -mindepth 1 -maxdepth 1 -type d` &&
   kim-api-collections-management install system `find ../examples/simulator-models -mindepth 1 -maxdepth 1 -type d` &&
   kim-api-collections-management list &&
-  ./test-install-relocated/libexec/kim-api/kim-api-simulator-model Sim_LAMMPS_LJcut_AkersonElliott_Alchemy_PbAu smspec-file data \
+  kim-api-simulator-model Sim_LAMMPS_LJcut_AkersonElliott_Alchemy_PbAu smspec-file data \
```
Comment (ellio167) [inline scripts/ci-build-and-test:123]: This won't work on linux/mac since `.../libexec/...` is not on the path

File: utils/CMakeLists.txt [hunk 5]
```diff
@@ -111,6 +121,7 @@ target_include_directories(portable-model-info PRIVATE "${PROJECT_SOURCE_DIR}/cp
 set_target_properties(portable-model-info PROPERTIES OUTPUT_NAME ${PORTABLE_MODEL_INFO})
 
 add_executable(simulator-model simulator-model.cpp)
+target_include_directories(simulator-model PRIVATE "${PROJECT_SOURCE_DIR}/cpp/src")
```
Comment (ellio167) [inline utils/CMakeLists.txt:124]: Why is this necessary?

File: scripts/ci-build-and-test (hunk 2 above)
Comment (stukowski) [reply scripts/ci-build-and-test:80]: In the MinGW environment, CMake seems to automatically adjust the `CMAKE_INSTALL_PREFIX`. Even though CMake is being called as follows:

```
cmake -DCMAKE_INSTALL_PREFIX=${INSTALL_PREFIX:=/usr/local} ...
```

the effective value of `CMAKE_INSTALL_PREFIX` turns out to be `C:/tools/msys64/usr/local`. You can observe this in the [log output](https://travis-ci.com/github/stukowski/kim-api/builds/200086002#L606).

The purpose of the code I've added is to query the actual value of the `CMAKE_INSTALL_PREFIX` variable and use in the following lines to compute the install path. It's not strictly necessary to (re-)use the shell variable name "`INSTALL_PREFIX`" for this. But I thought it makes sense. On Unix/Linux systems, it will already have the same value as `CMAKE_INSTALL_PREFIX` and overwriting it shouldn't hurt.

File: scripts/ci-build-and-test (hunk 4 above)
Comment (stukowski) [reply scripts/ci-build-and-test:123]: Then we might have to use if/else for conditional execution of this statement on Windows. On the Windows platform, the executable `kim-api-simulator-model` gets installed in `bin/`, not `libexec/kim-api/`.

File: utils/CMakeLists.txt (hunk 5 above)
Comment (stukowski) [reply utils/CMakeLists.txt:124]: I'm glad you asked. In fact, it's no longer needed. I introduced it, because I had started to replace some of the path handling code in `simulator-model.cpp` with our new `KIM::FILESYSTEM::Path` class in order to solve a problem, which I thought was caused by these kinds of statements:
```
std::string const filePath = *dirName + "/" + *name;
```
I thought it would be necessary to replace them with a more portable solution and made sure that the code has access to `KIM_FilesystemPath.hpp`. 

However, I've reversed these code changes after I realized that the problem was not related to file path handling. Instead, it was caused by the different line endings on the Windows platform. That's why you see a different change to `simulator-model.cpp` in my updated pull request: The `smspec` file must be opened in binary mode ("rb") to make sure the number of bytes read from the file exactly matches the file's size. 

I will revert the change made to the CMakeLists.txt file.

File: examples/simulators/utility_forces_numer_deriv/CMakeLists.txt (hunk 3 above)
Comment (stukowski) [reply examples/simulators/utility_forces_numer_deriv/CMakeLists.txt:38]: I cannot explain why, but without this change execution of the commands from `INSTALL` fails with the following error:
```
...
++ cp -r /home/astuk/kim-api/build/../examples/simulators/utility_forces_numer_deriv /home/astuk/kim-api/build/WORKSPACE/
++ cd /home/astuk/kim-api/build/WORKSPACE
++ cd utility_forces_numer_deriv
++ mkdir build
++ cd build
++ cmake ..
-- Building for: MinGW Makefiles
-- The Fortran compiler identification is GNU 10.1.0
-- Check for working Fortran compiler: C:/msys64/mingw64/bin/gfortran.exe
-- Check for working Fortran compiler: C:/msys64/mingw64/bin/gfortran.exe - works
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Checking whether C:/msys64/mingw64/bin/gfortran.exe supports Fortran 90
-- Checking whether C:/msys64/mingw64/bin/gfortran.exe supports Fortran 90 - yes
-- Configuring done
CMake Error in CMakeLists.txt:
  Cannot use features from non-enabled language CXX

CMake Generate step failed.  Build files cannot be regenerated correctly.
``` 
Adding "CXX" to the language list helped overcome the problem.

Comment (stukowski) [reply examples/simulators/utility_forces_numer_deriv/CMakeLists.txt:38]: CMake version is 3.17.3, by the way.

**Comment by stukowski** [comment]
> Once in a while the download fails for some reason. I would re-trigger the ci build and see if it succeeds.
> 
> Great that you have it all working!

Found out that it's not the download that hangs but the following `tar` command:

```
+ printf 'Downloading..............#%s\n' EDIP__MD_506186535567_002
+ sed -e 's/ /./g' -e 's/#/ /g'
Downloading.............. EDIP__MD_506186535567_002
+ wget --user-agent 'kim-api-collections-management-(2.2.0-git+v2.1.3-git-207-g5e4ec075.dirty.GNU.GNU.GNU--MINGW64_NT-10.0-19041-3.1.6-340.x86_64-x86_64)' -q --content-disposition https://openkim.org/download/EDIP__MD_506186535567_002.txz
+ local error_code=0
+ test 0 -eq 0
+ tar Jxf EDIP__MD_506186535567_002.txz
```

Seems to be Windows Antimalware service interfering, see https://github.com/msys2/MSYS2-packages/issues/1548.
I've implemented this [workaround](https://github.com/openkim/kim-api/pull/56/commits/0736e6b2d260da4443d4ad12efb48db62f547cd1):

```
xz --decompress --stdout "${archive_basename}.txz" | tar xf -
```

File: examples/simulators/utility_forces_numer_deriv/CMakeLists.txt (hunk 3 above)
Comment (ellio167) [reply examples/simulators/utility_forces_numer_deriv/CMakeLists.txt:38]: I think this is caused by the compile features setting:
https://github.com/openkim/kim-api/blob/491242cb0af43aa1a3b3dbcefbfce07807adce8a/cpp/src/CMakeLists.txt#L89-L100

Any idea what the "right" way to handle this situation is?  I wonder if the CMake way would be to create separate library targets (alias targets?) for each language binding?

Comment (stukowski) [reply examples/simulators/utility_forces_numer_deriv/CMakeLists.txt:38]: Apparently, the CMake developers are already aware of the problem and have fixed it in CMake 3.18.0 (too late for us, unfortunately):

https://gitlab.kitware.com/cmake/cmake/-/issues/20438

With the new CMake version, the consuming target (Fortran-only) will ignore any C++-related requirements such as `cxx_std_17` of the consumed target (kim-api) if the C++ language has not been explicitly enabled. 

To tell the truth, I'm not sure either how to deal with the problem in earlier CMake versions. The `kim-api-targets.cmake` file is set in stone once it has been generated and cannot possibly know which languages are enabled by the consuming target. Even injecting a call to [`enable_language(CXX)`](https://cmake.org/cmake/help/latest/command/enable_language.html) into `kim-api-targets.cmake` appears futile because of the scoping rules it is subject to.

I think, for the time being, developers of consuming targets are forced to explicitly enable `CXX` unless they are using CMake 3.18 or above.

Comment (ellio167) [reply examples/simulators/utility_forces_numer_deriv/CMakeLists.txt:38]: > Apparently, the CMake developers are already aware of the problem and have fixed it in CMake 3.18.0 (too late for us, unfortunately):
> 
> https://gitlab.kitware.com/cmake/cmake/-/issues/20438
> 
> With the new CMake version, the consuming target (Fortran-only) will ignore any C++-related requirements such as cxx_std_17 of the consumed target (kim-api) if the C++ language has not been explicitly enabled.

Nice find!  Thanks.  However, the fix was NOT merged into 3.18.0 due to the identification of a troublesome corner case.  As of now it seems the issue is still an open problem.

It seems to me the only way to deal with this at the moment would be to define multiple "interface specific" kim-api targets and export each one for use by consuming projects.  However, this seems like it might be overkill; maybe.  

(The worst case scenario, I think, is something like a big Fortran package that supports optional use of the KIM API.  Either they would need to include special logic to only link with KIM-API if specifically requested; Or their build would explicitly depend on a C++ environment being available.  That could be a deal-breaker; certainly I would not be happy in that situation.)

I'll think on this a bit, but probably we'll go with things as you have them.

Comment (ellio167) [reply examples/simulators/utility_forces_numer_deriv/CMakeLists.txt:38]: OK; Yes, let's go with it as is except for a minor adjustment which I'll push here nex.
